validator_client/src/duties_service.rs

```diff
@@ -1,758 +1,755 @@
-
-                validator_map.get(&epoch).and_then(|duties| {
-                    if duties.duty.attestation_slot == Some(slot) {
-                        Some(duties)
-                    } else {
-                        None
-                    }
-                })
-            })
+            .filter_map(|(_, map)| map.get(&epoch))
+            .map(|(_, duty_and_proof)| duty_and_proof)
+            .filter(|duty_and_proof| duty_and_proof.duty.slot == slot)
             .cloned()
             .collect()
     }
+}
 
-    fn get_index(&self, pubkey: &PublicKey, epoch: Epoch) -> Option<u64> {
-        self.store
-            .read()
-            .get(pubkey)?
-            .get(&epoch)?
-            .validator_index()
-    }
-
-    fn is_aggregator(&self, validator_pubkey: &PublicKey, epoch: Epoch) -> Option<bool> {
-        Some(
-            self.store
-                .read()
-                .get(validator_pubkey)?
-                .get(&epoch)?
-                .selection_proof
-                .is_some(),
-        )
-    }
-
-    fn insert<T: SlotClock + 'static, E: EthSpec>(
-        &self,
-        epoch: Epoch,
-        mut duties: DutyAndProof,
-        slots_per_epoch: u64,
-        validator_store: &ValidatorStore<T, E>,
-        spec: &ChainSpec,
-    ) -> Result<InsertOutcome, String> {
-        let mut store = self.store.write();
-
-        if !duties_match_epoch(&duties.duty, epoch, slots_per_epoch) {
-            return Ok(InsertOutcome::Invalid);
-        }
+/// Start the service that periodically polls the beacon node for validator duties. This will start
+/// several sub-services.
+///
+/// ## Notes
+///
+/// The loops in this function are structured such that a new instance of that task will only start
+/// once the current one is finished. This means that if a task happens to take more than one slot
+/// to run, we might skip a slot. This is unfortunate, however the alternative is to *always*
+/// process every slot, which has the chance of creating a theoretically unlimited backlog of tasks.
+/// It was a conscious decision to choose to drop tasks on an overloaded/latent system rather than
+/// overload it even more.
+pub fn start_update_service<T: SlotClock + 'static, E: EthSpec>(
+    core_duties_service: Arc<DutiesService<T, E>>,
+    mut block_service_tx: Sender<BlockServiceNotification>,
+) {
+    /*
+     * Spawn the task which updates the map of pubkey to validator index.
+     */
+    let duties_service = core_duties_service.clone();
+    core_duties_service.context.executor.spawn(
+        async move {
+            loop {
+                match duties_service.slot_clock.duration_to_next_slot() {
+                    Some(duration) => {
+                        // Run this poll before the wait, this should hopefully download all the indices
+                        // before the block/attestation tasks need them.
+                        poll_validator_indices(&duties_service).await;
```

Comment:
I don't know if it's just because I've been thinking about clock lag too much recently, but I wonder if we risk misalignment issues by getting the duration, doing some work, then waiting the duration? What if we put the poll before the `match` instead?

Reply:
Good call!